File: src/api/FileSystem.ts
import Path from 'path'
import FS from 'fs'

export class ManilaFile {
	constructor(...f: string[]) {
		const path = Path.join(...f)
		if (Path.isAbsolute(path)) this.#path = path
		else this.#path = Path.join(process.cwd(), path)

		if (FS.existsSync(this.#path) && FS.lstatSync(this.#path).isDirectory())
			throw new Error(`Path '${this.#path}' is a directory, not a file!`)
	}

	/**
	 * Checks if the file exists
	 */
	exists(): boolean {
		return FS.existsSync(this.#path)
	}
	/**
	 * Write data into a file
	 * @param data the data to write
	 * @param encoding the encoding the file will be saved in
	 */
	write(data: string | Buffer, encoding: BufferEncoding | null | undefined = 'utf-8'): ManilaFile {
		FS.writeFileSync(this.#path, data, { encoding })
		return this
	}
	/**
	 * Read data from a file
	 * @param encoding the encoding of the file
	 */
	read(encoding: BufferEncoding | null | undefined = 'utf-8'): string | Buffer {
		return FS.readFileSync(this.#path, { encoding: encoding })
	}

	/**
	 * Deletes the file (not this handle)
	 */
	delete(): ManilaFile {
		FS.rmSync(this.#path)
		return this
	}
	/**
	 * Copies the file into another location
	 * @param to the destination directory
	 */
	copy(...to: string[]): ManilaFile {
		FS.copyFileSync(this.#path, Path.join(...to))
		return this
	}
	/**
	 * Moves or renames the file
	 * @param to the destination
	 */
	move(...to: string[]): ManilaFile {
		FS.renameSync(this.#path, Path.join(...to))
		this.#path = Path.join(...to)
		return this
	}

	/**
	 * Gets the full path of the file
	 */
	getPath(): string {
		return this.#path
	}
	/**
	 * Gets the relative path to this file from a given origin
	 * @param from origin of your path
	 */
	getPathRelative(from: string): string {
		return Path.relative(from, this.#path)
	}
	/**
	 * Gets the directory this file is placed in
	 */
	getDir(): string {
		return Path.dirname(this.#path)
	}
	/**
	 * Gets the name of the file
	 */
	getFileName(): string {
		return Path.basename(this.#path)
	}

	#path: string
}

export class ManilaDirectory {
	constructor(...dir: string[]) {
		const path = Path.join(...dir)
		if (Path.isAbsolute(path)) this.#path = path
		else this.#path = Path.join(process.cwd(), path)

		if (FS.existsSync(this.#path) && FS.lstatSync(this.#path).isFile())
			throw new Error(`Path '${this.#path}' is a file, not a directory!`)
	}

	/**
	 * Checks if the directory exists
	 */
	exists(): boolean {
		return FS.existsSync(this.#path)
	}
	/**
	 * Creates the directory
	 * @param recursive create sub directories
	 */
	create(recursive: boolean = true): ManilaDirectory {
		FS.mkdirSync(this.#path, { recursive })
		return this
	}
	/**
	 * Deletes the directory
	 * @param recursive delete sub directories
	 */
	delete(recursive: boolean = true): ManilaDirectory {
		FS.rmSync(this.#path, { recursive })
		return this
	}
	/**
	 * Renames the directory basename
	 * @param to the new name
	 */
	rename(to: string): ManilaDirectory {
		FS.renameSync(this.#path, Path.join(Path.dirname(this.#path), to))
		this.#path = Path.join(Path.dirname(this.#path), to)
		return this
	}
	/**
	 * Moves the directory into another location
	 * @param to the destination
	 */
	move(to: string): ManilaDirectory {
		FS.renameSync(this.#path, Path.join(process.cwd(), to, Path.basename(this.#path)))
		this.#path = Path.join(process.cwd(), to, Path.basename(this.#path))
		return this
	}
	/**
	 * Concats a path to the current directory
	 * @param paths the paths to conact
	 */
	concat(...paths: string[]) {
		this.#path = Path.join(this.#path, ...paths)
		return this
	}
	/**
	 * Gets the files contained in the directory
	 * @param recursive also searches in sub-directories
	 */
	files(recursive: boolean = false, predicate: (f: string) => boolean): ManilaFile[] {
		const out: ManilaFile[] = []
		if (!recursive) {
			for (const f of FS.readdirSync(this.#path)) {
				out.push(new ManilaFile(this.#path, f))
			}
			return out
		}

		function rec(root: string, path: string) {
			const files = FS.readdirSync(path)
			for (const file of files) {
				const dir = Path.join(path, file)
				const stat = FS.lstatSync(dir)
				if (stat.isDirectory()) rec(root, dir)
				else if (stat.isFile()) {
					if (predicate != undefined) {
						if (predicate(Path.relative(root, dir))) out.push(new ManilaFile(dir))
					} else {
						out.push(new ManilaFile(dir))
					}
				}
			}
		}

		rec(this.#path, this.#path)
		return out
	}

	/**
	 * Returns the full path of the directory
	 */
	getPath() {
		return this.#path
	}
	/**
	 * Gets the relative path to this directory from a given origin
	 * @param from origin of your path
	 */
	getPathRelative(from: string): string {
		return Path.relative(from, this.#path)
	}
	/**
	 * Gets the last directory name (basename)
	 */
	getBasename() {
		return Path.basename(this.#path)
	}

	#path: string
}

File: Manila/src/scripting/api/FileSystem.cs

using Newtonsoft.Json;
using Microsoft.ClearScript;

namespace Manila.Scripting.API;

/// <summary>
/// A file handle
/// </summary>
public class ManilaFile {
	/// <summary>
	/// The absolute path this handle is pointing to
	/// </summary>
	public string path { get; }

	public ManilaFile(params string[] path) {
		if (path.Length < 1) throw new ArgumentException("Argument 'path' must contain at least one valid path!");
		var inPath = Path.IsPathRooted(path[0]) ? "" : Directory.GetCurrentDirectory();
		foreach (var p in path) {
			inPath = Path.Join(inPath, p);
		}

		if (Path.IsPathRooted(inPath)) this.path = inPath;
		else this.path = Path.Join(Directory.GetCurrentDirectory(), inPath);

		if (File.Exists(this.path) && File.GetAttributes(this.path) == FileAttributes.Directory) throw new ArgumentException("The path '" + this.path + "' is a directory!");
	}
	public ManilaFile(ManilaDirectory dir, string file) {
		path = Path.Join(dir.getPath(), file);

	}

	/// <summary>
	/// Writes a string to a file
	/// </summary>
	public ManilaFile write(string data) {
		string dir = Path.GetDirectoryName(path);
		if (!Directory.Exists(dir)) Directory.CreateDirectory(dir);
		File.WriteAllText(path, data);
		return this;
	}
	/// <summary>
	/// Writes a string array to a file
	/// </summary>
	public ManilaFile write(params string[] data) {
		string dir = Path.GetDirectoryName(path);
		if (!Directory.Exists(dir)) Directory.CreateDirectory(dir);
		File.WriteAllText(path, string.Join("\n", data));
		return this;
	}
	/// <summary>
	/// Writes a dynamic type to a file
	/// </summary>
	public ManilaFile write(dynamic data) {
		string dir = Path.GetDirectoryName(path);
		if (!Directory.Exists(dir)) Directory.CreateDirectory(dir);
		File.WriteAllText(path, string.Join("\n", data));
		return this;
	}

	/// <summary>
	/// Deletes a file (not the handle)
	/// </summary>
	public ManilaFile delete() {
		string dir = Path.GetDirectoryName(path);
		if (!Directory.Exists(dir)) Directory.CreateDirectory(dir);
		if (File.Exists(path)) File.Delete(path);
		return this;
	}

	/// <summary>
	/// Checks if a file exists
	/// </summary>
	public bool exists() {
		return File.Exists(path);
	}
	/// <summary>
	/// Reads the file as a string
	/// </summary>
	public string read() {
		return File.ReadAllText(path);
	}

	/// <summary>
	/// Deserializes a json file to an object
	/// </summary>
	/// <typeparam name="T">the type of the object</typeparam>
	/// <returns>the value of the object</returns>
	/// <exception cref="FileNotFoundException"></exception>
	public T deserializeJSON<T>() where T : new() {
		if (!exists()) throw new FileNotFoundException("File '" + getPath() + "' does not exist!");
		T? res = JsonConvert.DeserializeObject<T>(read());
		if (res == null) return new T();
		return res;
	}
	/// <summary>
	/// Serializes an object as a JSON file
	/// </summary>
	/// <param name="data">the object to serialize</param>
	/// <param name="format">if the json file should be formatted</param>
	public ManilaFile serializeJSON(object data, bool format = false) {
		write(JsonConvert.SerializeObject(data, format ? Formatting.Indented : Formatting.None));
		return this;
	}

	/// <summary>
	/// Renames a file
	/// </summary>
	/// <param name="name">new name</param>
	public ManilaFile rename(string name) {
		File.Move(path, name);
		return new ManilaFile(Path.Join(Directory.GetCurrentDirectory(), name));
	}
	/// <summary>
	/// Moves a file
	/// </summary>
	/// <param name="destination">the destination</param>
	public ManilaFile move(string destination) {
		if (Path.IsPathFullyQualified(destination)) {
			if (!Directory.Exists(destination)) Directory.CreateDirectory(destination);
			File.Move(path, destination);
			return this;
		}

		string newPath = Path.Join(Directory.GetCurrentDirectory(), destination, Path.GetFileName(path));
		if (!Directory.Exists(Path.GetDirectoryName(newPath))) Directory.CreateDirectory(Path.GetDirectoryName(newPath));
		File.Move(path, newPath);
		return new ManilaFile(newPath);
	}
	/// <summary>
	/// Copies a file to a new destinaton
	/// </summary>
	/// <param name="destination">the copy destination</param>
	/// <returns>the unchanged file handle</returns>
	public ManilaFile copy(string destination) {
		if (Path.IsPathRooted(destination)) {
			if (Directory.Exists(destination) && File.GetAttributes(destination) == FileAttributes.Directory) {
				File.Copy(path, Path.Join(destination, Path.GetFileName(path)));
			} else {
				File.Copy(path, destination, true);
			}
		} else {
			File.Copy(path, Path.Join(Directory.GetCurrentDirectory(), destination, Path.GetFileName(path)));
		}
		return this;
	}

	/// <summary>
	/// Gets the full parsed file path (will remove navigators like ../)
	/// </summary>
	public string getPath() { return Path.GetFullPath(path); }
	/// <summary>
	/// Gets the relative path from the current path to the specified destination path
	/// </summary>
	/// <param name="to">the destination path (default is current directory)</param>
	public string getPathRelative(string to = ".") { return Path.GetRelativePath(to, path); }
	/// <summary>
	/// Gets the name of the targeted file
	/// </summary>
	public string getFileName() { return Path.GetFileName(path); }
	/// <summary>
	/// Gets the path where the file is located in
	/// </summary>
	public string getFileDir() { return Path.GetDirectoryName(path); }
	public ManilaDirectory getFileDirHandle() { return new ManilaDirectory(getFileDir()); }

	public ManilaFile setExtension(string e) {
		return new ManilaFile(Path.Join(Path.GetDirectoryName(path), $"{Path.GetFileNameWithoutExtension(path)}.{e}"));
	}
}

/// <summary>
/// A directory hanlde
/// </summary>
public class ManilaDirectory {
	/// <summary>
	/// The absolute path this handle is pointing to
	/// </summary>
	public string path { get; }

	public ManilaDirectory(params string[] path) {
		if (path.Length < 1) throw new ArgumentException("Argument 'path' must contain at least one valid path!");
		var inPath = Path.IsPathRooted(path[0]) ? "" : Directory.GetCurrentDirectory();
		foreach (var p in path) {
			inPath = Path.Join(inPath, p);
		}

		this.path = inPath;
	}
	public ManilaDirectory(ManilaDirectory root, params string[] path) : this(Path.Join(root.getPath(), string.Join(Path.PathSeparator, path))) {
	}

	/// <summary>
	/// Joins a path to the current path
	/// </summary>
	/// <param name="paths">the paths to join</param>
	public ManilaDirectory join(params string[] paths) {
		var temp = path;
		foreach (var p in paths) {
			temp = Path.Join(temp, p);
		}
		return new ManilaDirectory(temp);
	}

	/// <summary>
	/// Gets the list of files contained in a directory
	/// </summary>
	/// <param name="func">the function to filter the files files</param>
	/// <param name="deep">flag to indicate whether subdirectories should also be searched</param>
	public ManilaFile[] files(ScriptObject? func = null, bool deep = false) {
		List<ManilaFile> result = new List<ManilaFile>();

		if (deep) {
			var files = Directory.GetFiles(path, "*", SearchOption.AllDirectories);
			foreach (var file in files) {
				if (func == null || (bool) func.InvokeAsFunction(file)) {
					result.Add(Manila.file(file));
				}
			}
		} else {
			var files = Directory.GetFiles(path);
			foreach (var file in files) {
				if (func == null || (bool) func.InvokeAsFunction(file)) {
					result.Add(Manila.file(file));
				}
			}
		}

		return result.ToArray();
	}

	/// <summary>
	/// Gets the list of files contained in a directory
	/// </summary>
	/// <param name="func">the function to filter the files files</param>
	/// <param name="deep">flag to indicate whether subdirectories should also be searched</param>
	public List<ManilaFile> files(bool deep, Func<string, bool> func) {
		List<ManilaFile> result = new List<ManilaFile>();

		if (deep) {
			var files = Directory.GetFiles(path, "*", SearchOption.AllDirectories);
			foreach (var file in files) {
				if (func == null || func(file)) {
					result.Add(Manila.file(file));
				}
			}
		} else {
			var files = Directory.GetFiles(path);
			foreach (var file in files) {
				if (func == null || func(file)) {
					result.Add(Manila.file(file));
				}
			}
		}

		return result;
	}

	/// <summary>
	/// Gets the list of files contained in a directory
	/// </summary>
	/// <param name="func">the function to filter the files files</param>
	/// <param name="deep">flag to indicate whether subdirectories should also be searched</param>
	public ManilaFile[] files(bool deep) {
		return files(null, deep);
	}

	/// <summary>
	/// Checks if the directory exists
	/// </summary>
	public bool exists() { return Directory.Exists(path); }

	/// <summary>
	/// Creates the directory if it does not exist
	/// </summary>
	public void create() { if (!exists()) Directory.CreateDirectory(path); }

	/// <summary>
	/// Gets the full parsed path (will remove navigators like ../)
	/// </summary>
	public string getPath() { return Path.GetFullPath(path); }
	/// <summary>
	/// Gets the relative path from the current path to the specified destination path
	/// </summary>
	/// <param name="to">the destination path (default is current directory)</param>
	public string getPathRelative(string to = ".") { return Path.GetRelativePath(to, path); }

	/// <summary>
	/// Gets a ManilaFile handle to the specified file
	/// </summary>
	/// <param name="files">files to search</param>
	/// <returns>the file handle from the found file (if none is found, it retruns the first)</returns>
	public ManilaFile file(params string[] files) {
		if (files.Length < 1) throw new ArgumentException("Argument 'files' requires at least one filename!");

		foreach (var f in files) {
			ManilaFile handle = Manila.file(getPath(), f);
			if (handle.exists()) return handle;
		}
		return Manila.file(getPath(), files[0]);
	}
}
